File: parser.rb
# parser.rb
require 'net/http'
require 'json'
require 'csv'
require 'time'

class RateParser
  API_URL = 'https://api.exchangerate.host/latest?base=USD'
  CACHE_TTL = 60

  def initialize
    @rates = {}
    @last_update = nil
    @history = []
    fetch_rates
  end

  def fetch_rates
    uri = URI(API_URL)
    begin
      response = Net::HTTP.get_response(uri)
      if response.is_a?(Net::HTTPSuccess)
        data = JSON.parse(response.body)
        if data['success']
          @rates = data['rates']
          @last_update = Time.now
          @history << { timestamp: @last_update.iso8601, rates: @rates.dup }
          @history.shift if @history.size > 10
          return true
        end
      end
      false
    rescue => e
      puts "Error fetching rates: #{e.message}"
      false
    end
  end

  def get_rates
    if @last_update && (Time.now - @last_update) > CACHE_TTL
      fetch_rates
    end
    @rates
  end

  def filter_rates(currencies)
    all_rates = get_rates
    all_rates.select { |k, _| currencies.include?(k) }
  end

  def search_currency(query)
    all_rates = get_rates
    q = query.downcase
    all_rates.select { |k, _| k.downcase.include?(q) }
  end

  def refresh
    fetch_rates
  end

  def export_json(filename = 'rates.json')
    data = {
      timestamp: @last_update&.iso8601,
      rates: get_rates
    }
    File.write(filename, JSON.pretty_generate(data))
    true
  rescue => e
    puts "Export failed: #{e.message}"
    false
  end

  def export_csv(filename = 'rates.csv')
    rates = get_rates
    CSV.open(filename, 'w') do |csv|
      csv << ['Currency', 'Rate']
      rates.keys.sort.each do |k|
        csv << [k, rates[k].round(4)]
      end
    end
    true
  rescue => e
    puts "Export failed: #{e.message}"
    false
  end

  def show_history
    if @history.empty?
      puts 'No history yet.'
      return
    end
    @history.each_with_index do |entry, i|
      puts "[#{i+1}] #{entry[:timestamp]} – #{entry[:rates].size} currencies"
    end
  end

  def display_rates(rates = nil)
    rates ||= get_rates
    if rates.empty?
      puts 'No rates available.'
      return
    end
    puts "\nRates (USD base) – updated: #{@last_update || 'unknown'}"
    puts '-' * 40
    rates.keys.sort.first(20).each do |k|
      puts "#{k.ljust(5)} : #{rates[k].round(4)}"
    end
    if rates.size > 20
      puts "... and #{rates.size - 20} more"
    end
  end
end

def main
  parser = RateParser.new
  puts "=== Exchange Rate Parser ==="
  loop do
    puts "\n1. Show all rates"
    puts "2. Filter by currency"
    puts "3. Search currency"
    puts "4. Export to JSON"
    puts "5. Export to CSV"
    puts "6. Refresh rates"
    puts "7. Show history"
    puts "8. Exit"
    print "Choose: "
    choice = gets.chomp.strip
    case choice
    when '1'
      parser.display_rates
    when '2'
      print "Enter currency codes (comma-separated): "
      input = gets.chomp.strip.upcase
      currencies = input.split(',').map(&:strip).reject(&:empty?)
      if currencies.any?
        filtered = parser.filter_rates(currencies)
        parser.display_rates(filtered)
      else
        puts "No currencies specified."
      end
    when '3'
      print "Enter currency code or name: "
      query = gets.chomp.strip
      if !query.empty?
        found = parser.search_currency(query)
        parser.display_rates(found)
      else
        puts "Query cannot be empty."
      end
    when '4'
      print "Filename (default: rates.json): "
      fname = gets.chomp.strip
      fname = 'rates.json' if fname.empty?
      if parser.export_json(fname)
        puts "Exported to #{fname}"
      end
    when '5'
      print "Filename (default: rates.csv): "
      fname = gets.chomp.strip
      fname = 'rates.csv' if fname.empty?
      if parser.export_csv(fname)
        puts "Exported to #{fname}"
      end
    when '6'
      if parser.refresh
        puts "Rates refreshed."
      else
        puts "Refresh failed."
      end
    when '7'
      parser.show_history
    when '8'
      puts "Goodbye!"
      break
    else
      puts "Invalid choice."
    end
  end
end

main if __FILE__ == $0

File: parser.py
# parser.py
import requests
import json
import csv
import time
from datetime import datetime
from typing import Dict, List, Optional

class RateParser:
    def __init__(self, cache_ttl: int = 60):
        self.api_url = "https://api.exchangerate.host/latest?base=USD"
        self.rates: Dict[str, float] = {}
        self.last_update: Optional[datetime] = None
        self.cache_ttl = cache_ttl
        self.history: List[Dict] = []
        self._fetch_rates()

    def _fetch_rates(self) -> bool:
        try:
            response = requests.get(self.api_url, timeout=5)
            if response.status_code == 200:
                data = response.json()
                if data.get("success"):
                    self.rates = data["rates"]
                    self.last_update = datetime.now()
                    self.history.append({
                        "timestamp": self.last_update.isoformat(),
                        "rates": self.rates.copy()
                    })
                    if len(self.history) > 10:
                        self.history.pop(0)
                    return True
            return False
        except Exception as e:
            print(f"Error fetching rates: {e}")
            return False

    def get_rates(self) -> Dict[str, float]:
        if self.rates and self.last_update:
            if (datetime.now() - self.last_update).seconds > self.cache_ttl:
                self._fetch_rates()
        return self.rates

    def filter_rates(self, currencies: List[str]) -> Dict[str, float]:
        rates = self.get_rates()
        return {k: v for k, v in rates.items() if k in currencies}

    def search_currency(self, query: str) -> Dict[str, float]:
        rates = self.get_rates()
        query_lower = query.lower()
        return {k: v for k, v in rates.items() if query_lower in k.lower()}

    def refresh(self) -> bool:
        return self._fetch_rates()

    def export_json(self, filename: str = "rates.json") -> bool:
        try:
            data = {
                "timestamp": self.last_update.isoformat() if self.last_update else None,
                "rates": self.get_rates()
            }
            with open(filename, 'w') as f:
                json.dump(data, f, indent=2)
            return True
        except Exception as e:
            print(f"Export failed: {e}")
            return False

    def export_csv(self, filename: str = "rates.csv") -> bool:
        try:
            rates = self.get_rates()
            with open(filename, 'w', newline='') as f:
                writer = csv.writer(f)
                writer.writerow(["Currency", "Rate"])
                for curr, rate in sorted(rates.items()):
                    writer.writerow([curr, rate])
            return True
        except Exception as e:
            print(f"Export failed: {e}")
            return False

    def show_history(self):
        if not self.history:
            print("No history yet.")
            return
        for i, entry in enumerate(self.history[-5:], 1):
            print(f"[{i}] {entry['timestamp']} – {len(entry['rates'])} currencies")

    def display_rates(self, rates: Dict[str, float] = None):
        if rates is None:
            rates = self.get_rates()
        if not rates:
            print("No rates available.")
            return
        print(f"\nRates (USD base) – updated: {self.last_update}")
        print("-" * 40)
        for curr, rate in sorted(rates.items())[:20]:
            print(f"{curr:5} : {rate:.4f}")
        if len(rates) > 20:
            print(f"... and {len(rates)-20} more")

def main():
    parser = RateParser()
    print("=== Exchange Rate Parser ===")
    while True:
        print("\n1. Show all rates")
        print("2. Filter by currency")
        print("3. Search currency")
        print("4. Export to JSON")
        print("5. Export to CSV")
        print("6. Refresh rates")
        print("7. Show history")
        print("8. Exit")
        choice = input("Choose: ").strip()
        if choice == "1":
            parser.display_rates()
        elif choice == "2":
            currencies = input("Enter currency codes (comma-separated): ").strip().upper().split(',')
            currencies = [c.strip() for c in currencies if c.strip()]
            if currencies:
                filtered = parser.filter_rates(currencies)
                parser.display_rates(filtered)
            else:
                print("No currencies specified.")
        elif choice == "3":
            query = input("Enter currency code or name: ").strip()
            if query:
                found = parser.search_currency(query)
                parser.display_rates(found)
            else:
                print("Query cannot be empty.")
        elif choice == "4":
            filename = input("Filename (default: rates.json): ").strip() or "rates.json"
            if parser.export_json(filename):
                print(f"Exported to {filename}")
        elif choice == "5":
            filename = input("Filename (default: rates.csv): ").strip() or "rates.csv"
            if parser.export_csv(filename):
                print(f"Exported to {filename}")
        elif choice == "6":
            if parser.refresh():
                print("Rates refreshed.")
            else:
                print("Refresh failed.")
        elif choice == "7":
            parser.show_history()
        elif choice == "8":
            print("Goodbye!")
            break
        else:
            print("Invalid choice.")

if __name__ == "__main__":
    main()
